firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/Mutation.java

```diff
@@ -89,10 +170,120 @@ public Precondition getPrecondition() {
     return precondition;
   }
 
+  public MutationType getMutationType() {
+    return type;
+  }
+
+  public ObjectValue getValue() {
+    return value;
+  }
+
+  public FieldMask getMask() {
+    return mask;
+  }
+
   public List<FieldTransform> getFieldTransforms() {
     return fieldTransforms;
   }
 
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    Mutation that = (Mutation) o;
+    return this.type.equals(that.type)
+        && hasSameKeyAndPrecondition(that)
+        && hasSameObjectValue(that)
+        && hasSameFieldMask(that)
+        && hasSameTransforms(that);
+  }
+
+  @Override
+  public int hashCode() {
+    int hashCode = keyAndPreconditionHashCode();
+    hashCode = hashCode * 31 + type.hashCode();
+    hashCode = hashCode * 31 + valueHashCode();
+    hashCode = hashCode * 31 + maskHashCode();
+    hashCode = hashCode * 31 + transformsHashCode();
+
+    return hashCode;
+  }
+
+  @Override
+  public String toString() {
+    switch (type) {
+      case DELETE:
+        {
+          return "Mutation{ type=DELETE; " + keyAndPreconditionToString() + "}";
+        }
+        // TODO(Overlay): Add transforms here and below.
+      case SET:
+        {
+          return "Mutation{ type=SET; " + keyAndPreconditionToString() + ", value=" + value + "}";
+        }
+      case PATCH:
+        {
+          return "Mutation{ type=PATCH; "
+              + keyAndPreconditionToString()
+              + ", mask="
+              + mask
+              + ", value="
+              + value
+              + "}";
+        }
+      case VERIFY:
+        {
+          return "Mutation{ type=VERIRY; " + keyAndPreconditionToString();
+        }
+    }
+    hardAssert(false, "Unreachable");
```

Comment:
FWIW I think it would be perfectly fine to just use the toString() from Patch here and return null-values where applicable.